src/Compiler/Graphs/MinCut.cs

```diff
@@ -198,7 +200,11 @@ public Set<NodeType> GetSourceGroup()
                     foreach (EdgeType edge in graph.EdgesInto(node))
                     {
                         NodeType target = graph.SourceOf(edge);
-                        if (distanceToSink[node] != distanceToSink[target] + 1) continue;
+
+                        // Long to avoid overflow
+                        long nodeDistanceToSink = distanceToSink[node];
+                        long targetDistanceToSink = distanceToSink[target];
+                        if (nodeDistanceToSink != targetDistanceToSink + 1) continue;
                         // target cannot be in sourceGroup since their distances are huge,
                         // unless it is both a source and sink.
                         if (Sources.Contains(target)) continue;
```

Comment:
Moving this check earlier will prevent overflow above.  Also, it should be changed to `if (sourceGroup.Contains(target)) continue;`

Reply:
Added this check. And also changed the comparison from (x != y + 1) to (x - 1 != y) that removes the need to switch to long to avoid overflow.